robot-server/robot_server/lifetime_dependencies.py

```
@@ -0,0 +1,249 @@
+from robot_server.service.legacy.rpc import RPCServer
+from robot_server.service.protocol.manager import ProtocolManager
+from robot_server.service.session.manager import SessionManager
+
+
+_YieldT = typing.TypeVar("_YieldT")
+# Helper to be used as the return type annotation of factory functions that are
+# provided to contextlib.contextmanager.
+_CMFactory = typing.Generator[_YieldT, None, None]
+# Ditto, but for contextlib.asynccontextmanager.
+_ACMFactory = typing.AsyncGenerator[_YieldT, None]
+
+
+@contextlib.contextmanager
+def _prepared_event_publisher() -> _CMFactory[NotifyServerPublisher]:
+    notify_server_settings = NotifyServerSettings()
+    event_publisher = create_notify_server_publisher(
+        notify_server_settings.publisher_address.connection_string()
+    )
+    try:
+        yield event_publisher
+    finally:
+        event_publisher.close()
+
+
+@contextlib.asynccontextmanager
+async def _prepared_thread_manager(
+    event_publisher: NotifyServerPublisher,
+) -> _ACMFactory[SlowInitializing[ThreadManager]]:
+    async def factory() -> ThreadManager:
+        return await initialize_hardware(event_publisher)
+
+    slow_initializing = start_slow_initializing(factory)
+    try:
+        yield slow_initializing
+    finally:
+        try:
+            fully_initialized_result = await slow_initializing.get_when_ready()
+        except InitializationFailedError:
+            pass
+        else:
+            fully_initialized_result.clean_up()
+
+
+@contextlib.asynccontextmanager
+async def _prepared_rpc_server(
+    thread_manager: SlowInitializing[ThreadManager],
+    lock: ThreadedAsyncLock,
+) -> _ACMFactory[SlowInitializing[RPCServer]]:
+    async def factory() -> RPCServer:
+        root = APIMainRouter(hardware=await thread_manager.get_when_ready(), lock=lock)
+        return RPCServer(loop=None, root=root)
+
+    slow_initializing = start_slow_initializing(factory)
+    try:
+        yield slow_initializing
+    finally:
+        try:
```

This nested `try` might be hard to understand, and its duplication across all these context managers is a red flag. Centralize this logic under `SlowInitializing`, maybe.

---

Or:

```python
finally:
    with contextlib.suppress(InitializationFailedError):
        fully_initialized_result = slow_initializing.get_when_ready()
        await fully_initialized_result.on_shutdown()
```

Or:

Just not suppress the `InitializationFailedError`. Let it propagate?